packages/boot/test/unit/controller.booter.unit.ts

```diff
@@ -0,0 +1,400 @@
+        `${resolve(projectRoot, 'controllers/nested/nested.controller.js')}`,
+      ];
+
+      const booter = new ControllerBooter(app, bootOptions);
+      await booter.discover();
+      expect(booter.options.discovered).to.be.an.Array();
+      expect(booter.options.discovered.sort()).to.eql(expected.sort());
+    });
+
+    it('discovers files without going into nested folders', async () => {
+      const bootOptions: BootOptions = {
+        projectRoot: projectRoot,
+        controllers: Object.assign({}, ControllerDefaults, {
+          nested: false,
+        }),
+      };
+      const expected = [
+        `${resolve(projectRoot, 'controllers/empty.controller.js')}`,
+        `${resolve(projectRoot, 'controllers/hello.controller.js')}`,
+        `${resolve(projectRoot, 'controllers/two.controller.js')}`,
+      ];
+
+      const booter = new ControllerBooter(app, bootOptions);
+      await booter.discover();
+      expect(booter.options.discovered).to.be.an.Array();
+      expect(booter.options.discovered.sort()).to.eql(expected.sort());
+    });
+
+    it('discovers files of specified extensions', async () => {
+      const bootOptions: BootOptions = {
+        projectRoot: projectRoot,
+        controllers: Object.assign({}, ControllerDefaults, {
+          extensions: ['.ctrl.js'],
+        }),
+      };
+      const expected = [
+        `${resolve(projectRoot, 'controllers/another.ext.ctrl.js')}`,
+      ];
+
+      const booter = new ControllerBooter(app, bootOptions);
+      await booter.discover();
+      expect(booter.options.discovered).to.be.an.Array();
+      expect(booter.options.discovered.sort()).to.eql(expected.sort());
+    });
+
+    it('discovers files in specified directory', async () => {
+      const bootOptions: BootOptions = {
+        projectRoot: projectRoot,
+        controllers: Object.assign({}, ControllerDefaults, {
+          dirs: ['ctrl'],
+        }),
+      };
+      const expected = [
+        `${resolve(projectRoot, 'ctrl/multiple.folder.controller.js')}`,
+      ];
+
+      const booter = new ControllerBooter(app, bootOptions);
+      await booter.discover();
+      expect(booter.options.discovered).to.be.an.Array();
+      expect(booter.options.discovered.sort()).to.eql(expected.sort());
+    });
+
+    it('discovers files of multiple extensions', async () => {
+      const bootOptions: BootOptions = {
+        projectRoot: projectRoot,
+        controllers: Object.assign({}, ControllerDefaults, {
+          extensions: ['.ctrl.js', '.controller.js'],
+        }),
+      };
+      const expected = [
+        `${resolve(projectRoot, 'controllers/empty.controller.js')}`,
+        `${resolve(projectRoot, 'controllers/hello.controller.js')}`,
+        `${resolve(projectRoot, 'controllers/two.controller.js')}`,
+        `${resolve(projectRoot, 'controllers/another.ext.ctrl.js')}`,
+        `${resolve(projectRoot, 'controllers/nested/nested.controller.js')}`,
```

Comment:
As I mentioned before, having an external fixture with multiple artifacts is an anti-pattern that yields test suite that's difficult to maintain over time.

For example, in this particular test, the expected list is unnecessary repetitive. There is no need to specify all of `empty`, `hello` and `two` with the same extension `.controller.js`, one entry is enough to validate the logic.

A better solution is to create a set of helpers that make it super easy to prepare a test app directory with exactly the right (and minimal!) set of artifacts needed by the test. This makes it also much much easier to understand which parts of the fixture are relevant to the particular (usually failing) test.